webrender/src/prim_store.rs

```diff
@@ -0,0 +1,924 @@
+                    gpu_prim_index: gpu_address,
+                    gpu_data_address: GpuStoreAddress(0),
+                    gpu_data_count: 0,
+                };
+
+                self.cpu_images.push(image_cpu);
+                metadata
+            }
+            PrimitiveContainer::Border(border_cpu, border_gpu) => {
+                let gpu_address = self.gpu_data128.push(border_gpu);
+
+                let metadata = PrimitiveMetadata {
+                    is_opaque: false,
+                    need_to_build_cache: false,
+                    color_texture_id: TextureId(0),
+                    clip_index: clip_index,
+                    prim_kind: PrimitiveKind::Border,
+                    cpu_prim_index: SpecificPrimitiveIndex(self.cpu_borders.len()),
+                    gpu_prim_index: gpu_address,
+                    gpu_data_address: GpuStoreAddress(0),
+                    gpu_data_count: 0,
+                };
+
+                self.cpu_borders.push(border_cpu);
+                metadata
+            }
+            PrimitiveContainer::Gradient(gradient_cpu, gradient_gpu) => {
+                let gpu_address = self.gpu_data32.push(gradient_gpu);
+                let gpu_stops_address = self.gpu_data32.alloc(gradient_cpu.stops_range.length);
+
+                let metadata = PrimitiveMetadata {
+                    is_opaque: false,
+                    need_to_build_cache: true,
+                    color_texture_id: TextureId(0),
+                    clip_index: clip_index,
+                    prim_kind: PrimitiveKind::Gradient,
+                    cpu_prim_index: SpecificPrimitiveIndex(self.cpu_gradients.len()),
+                    gpu_prim_index: gpu_address,
+                    gpu_data_address: gpu_stops_address,
+                    gpu_data_count: gradient_cpu.stops_range.length as i32,
+                };
+
+                self.cpu_gradients.push(gradient_cpu);
+                metadata
+            }
+            PrimitiveContainer::BoxShadow(box_shadow, instance_rects) => {
+                let gpu_prim_address = self.gpu_data64.push(box_shadow);
+                let gpu_data_address = self.gpu_data16.get_next_address();
+
+                let metadata = PrimitiveMetadata {
+                    is_opaque: false,
+                    need_to_build_cache: false,
+                    color_texture_id: TextureId(0),
+                    clip_index: clip_index,
+                    prim_kind: PrimitiveKind::BoxShadow,
+                    cpu_prim_index: SpecificPrimitiveIndex::invalid(),
+                    gpu_prim_index: gpu_prim_address,
+                    gpu_data_address: gpu_data_address,
+                    gpu_data_count: instance_rects.len() as i32,
+                };
+
+                for rect in instance_rects {
+                    self.gpu_data16.push(InstanceRect {
+                        rect: rect,
+                    });
+                }
```

Comment:
nit: Could just be

    self.gpu_data16.extend(instance_rects.into_iter().map(|rect| InstanceRect {
        rect: rect,
    }))